### PR TITLE
feat(add): check default branch

### PR DESCRIPTION
**Pull request by carslen**
First attempt to implement check of default branch.

To get the setup default branch of a repository you need to query the GitHup API, which has some input parameters:

- Owner (aka GH Org)
- Repository name (e.g. tractusx-quality-checks)

The check needs to know these. When running as a GH Action/Workflow there are environment vars available, which makes it easy to retrieve these information. When running locally, it's a bit harder to extract these informations. In this first draft I decided to parse the `.git/config` file and use the `url` key.

For determining the repository name I've created the func `getRepoName()` which returns owner, repo(name) and a bool called local, which is true if `.git/config` was parsed.

A second func `getRepoInfo()` queries the GitHup API and returns `*github.Repository` object.

Finally, the func `QualityGuideline.Test()` first checks if it's executed on a fork and if true it returns the check as failed, due to unreliable circumstances. Then it tests for `DefaultBranch == main` and returns appropriate.

Currently proper testing is missing.

### CODE REVIEW

File: internal/default_branch.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+	if os.Getenv("GITHUB_REPOSITORY") != "" {
+		// env variable is available when executed as GH Action/Workflow/Check
+		owner = os.Getenv("GITHUB_REPOSITORY_OWNER")
+		repo = strings.Split(os.Getenv("GITHUB_REPOSITORY"), "/")[1]
+
+		return owner, repo, false
+	} else {
+		// Parse local git configuration when executing locally
+		cfg, err := ini.Load(".git/config")
```
Comment (almadigabor) [inline internal/default_branch.go:99]: You could also consider using git client and run `git remote get-url origin` but that would introduce a new dependency (if not already present).

File: internal/default_branch.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		SUFFIX  = ".git"
+	)
+
+	if os.Getenv("GITHUB_REPOSITORY") != "" {
+		// env variable is available when executed as GH Action/Workflow/Check
+		owner = os.Getenv("GITHUB_REPOSITORY_OWNER")
+		repo = strings.Split(os.Getenv("GITHUB_REPOSITORY"), "/")[1]
+
+		return owner, repo, false
+	} else {
```
Comment (SebastianBezold) [inline internal/default_branch.go:97]: No `else` necessary, if you have an early return on the first condition -> one less level of indentation

File: internal/default_branch.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		// Parse local git configuration when executing locally
+		cfg, err := ini.Load(".git/config")
+		if err != nil {
+			log.Fatalf("Failed to read file: %v", err)
```
Comment (SebastianBezold) [inline internal/default_branch.go:101]: Why `log` instead of `fmt`, like we did before. 
AFAIK, log would print to stderr by default, while fmt would print to stdout by default.
Not sure, if it is a big deal in our situation, but for consistency reasons, I would either stick with fmt, or change every output to use log. If we want to do it, we can do that in a separate PR/issue

File: internal/default_branch.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+func NewDefaultBranch() QualityGuideline {
+	return &defaultBranch{}
+}
+
+// getRepoName returns repoName (string) and local (bool). repoName is either
+// read from GITHUB_REPOSITORY environment variable when running as GH Action or,
+// when executed locally, parsing local Git config.
+//
+// local it true when repoName was parsed from local git config.
+func getRepoName() (owner string, repo string, local bool) {
```
Comment (SebastianBezold) [inline internal/default_branch.go:83]: Might be worth introducing a struct as return type. Functions with multiple return values of the same time tend to be less intuitive, since you need to check, which in which order the values are returned

**Review by SebastianBezold — COMMENTED** [review]
Would love to see tests for the repo config parsing 😁

File: internal/default_branch.go (hunk 1 above)
Comment (SebastianBezold) [reply internal/default_branch.go:99]: I know, that this might arguably be a low chance, that this would not work, but it would introduce a dependency to a binary, that has to be installed, and we need to take things like different OS and local setups into account, if we want that code to be runnable everywhere

Comment (Siegfriedk) [reply internal/default_branch.go:99]: There might be a go lang library for git support. I wouldn't might on a github based project to require git though

File: internal/default_branch.go (hunk 3 above)
Comment (almadigabor) [reply internal/default_branch.go:101]: Agree with the consistency. Just a question as I lack the knowledge: what would it result in sending the message to stderr instead of stdout?

File: internal/default_branch.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+func (d defaultBranch) Test() *QualityResult {
+	owner, repoName, isLocal := getRepoName()
+	repo, err := getRepoInfo(owner, repoName, isLocal)
+
+	if err != nil {
+		log.Fatalf("error %v", err)
+	}
+
+	if *repo.Fork {
+		return &QualityResult{Passed: false, ErrorDescription: "Check determined running on a fork."}
```
[the remark this answers is not in the document]
Comment (almadigabor) [reply internal/default_branch.go:57]: Setting `IsOptional()` on runtime is a weird behavior in my opinion. I would check for main default on forks as well.

File: internal/default_branch.go (hunk 1 above)
Comment (carslen) [reply internal/default_branch.go:99]: Of course there is a go git client library/package I could use to achieve the same. I decided to use the ini pkg. We could maybe compare which of the two packages might introduce less (indirect) dependencies to our repo and decide on that?

File: internal/default_branch.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (carslen) [reply internal/default_branch.go:57]: On forks Workflows are disabled by default in GitHub. I'll check if this might be the same for Checks.

[the remark this answers is not in the document]
Comment (carslen) [reply internal/default_branch.go:57]: OK, to test this, I have to open a PR to upstream.

File: internal/default_branch.go (hunk 1 above)
Comment (carslen) [reply internal/default_branch.go]: Package `go-ini/ini` has 1 indirect dependency, `go-git/go-git` has 18 indirect dependencies.

File: internal/default_branch.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+func (d defaultBranch) IsOptional() bool {
+	return false
+}
+
+func NewDefaultBranch() QualityGuideline {
+	return &defaultBranch{}
+}
+
+// getRepoName returns repoName (string) and local (bool). repoName is either
+// read from GITHUB_REPOSITORY environment variable when running as GH Action or,
```
[the remark this answers is not in the document]
Comment (carslen) [reply internal/default_branch.go:79]: Basically you're right. Environment var `GITHUB_REPOSITORY` could exists on a devs local machine. I searched for the easiest way to determine if running on GitHub, or not. But you're pointing out the godoc string here. I do have to rewrite the description anyways, I'll update it soon.

File: internal/default_branch.go (hunk 3 above)
Comment (carslen) [reply internal/default_branch.go]: Done.

File: internal/default_branch.go (hunk 4 above)
Comment (carslen) [reply internal/default_branch.go]: Done

File: internal/default_branch.go (hunk 1 above)
Comment (carslen) [reply internal/default_branch.go]: For now I would remain on pkg `go-ini`. If we later determine that `go-git` client is required for other checks as well, then let's refactor this.

**Review by almadigabor — APPROVED** [review]
LGTM
